### ISSUE TITLE
Apple Silicon M1 Support

### ISSUE THREAD

**Issue by n8henrie** [federico-terzi/espanso#615]
Thanks for your work on Espanso!

I don't see an existing issue tracking M1 Silicon support, so I thought I'd create one and report back.

Currently, installing Espanso via Homebrew installs an 86_64 version, even if using the `-s` (`--build-from-source`) option.

A `git clone` and `cargo build --release` works perfectly to create an arm64 version.

It looks like configuring to build a universal binary [should be pretty easy](https://developer.apple.com/documentation/apple-silicon/building-a-universal-macos-binary), though I've never done it before.

Unfortunately, it looks like the release artifacts are made with Azure Pipelines CI, which [doesn't yet have a MacOS 11 (Big Sur) option](https://github.com/actions/virtual-environments/issues/2072), so I'm not sure there is anything actionable on this yet.

But I thought it might be helpful for other users to know that a local clone and compilation works great on Apple Silicon, and hopefully once Azure supports Big Sure we can migrate the release artifact to a universal binary.

**Comment by federico-terzi** [comment]
Hey @n8henrie,

Thank you for the useful insights! Indeed, espanso does not "officially" support the M1 processors yet, though I'd love to do so in the upcoming months. I don't own a suitable mac at the moment, so I'll definitely need some help from the community if we want to support them :)

> even if using the `-s` (`--build-from-source`) option.

The current homebrew formula only supports pre-built binaries, but I'd love to support build-from-source mode. Do you have any experience with that?

At the moment, it seems that Rust does not support building universal binaries yet, though it might be coming: https://github.com/rust-lang/cargo/issues/8875

I'll post-pone this one after #594 is ready so that we have a solid foundation to support the new M1 processor

Let me know if you have any concerns

Cheers :)

**Comment by n8henrie** [comment]
Thanks for the very speedy response!

I've locally modified the homebrew formula to do a proper source installation and I'd be happy to make a PR.

However, it will first require making a point release to update the `Cargo.lock` due to [this issue](https://github.com/Homebrew/homebrew-core/pull/68089), without which the `cargo build` step fails.

Please note that in the below PR, I've updated the Cargo.lock so that it builds properly, but I did *not* increment the version (remains at 0.7.3).

**Comment by n8henrie** [comment]
The next step might be a little tricky -- I think the current build process makes for a recursive dependency problem with the formula, and I think the current formula may not be following homebrew best practice.

As I understand it, the formula *ought* to be building Espanso from source by default. It does so by downloading the compressed archive of a release, noting that it depends on `rust` and `cmake`, building and installing. At some point, a homebrew builder will (automatically?) create compiled artifacts known as "bottles" to allow speedier installation, which can then be included in the formula.

However, the current homebrew formula does essentially the opposite, by using the artifact and hash from an already compiled binary (which is why it installed `86_64` even when I tell it to build from source).

The problem in updating it to do things the proper way is that the archive contains the formula, so the hash of the archive is obviously going to change every time you update the formula with the hash. I'll see if I can figure out how others are dealing with this -- my past experience has been that the homebrew formula are not kept in the repo and are updated separately.

**Comment by n8henrie** [comment]
For context, looks like ripgrep has a separate [ripgrepbin formula](https://github.com/BurntSushi/ripgrep/blob/master/pkg/brew/ripgrep-bin.rb) that is kept in the repo and functions similarly to this one (no source install), but a different formula for source installs that is not in the repo.

[fd](https://github.com/sharkdp/fd/) also does not seem to have a formula in its repo.

**Comment by federico-terzi** [comment]
@n8henrie Thank you for the detailed explanation! Indeed, having a separate "espanso-bin" package might be a good solution in this case. I originally chose the "pre-built" route because I didn't want to force users to download the entire Rust toolchain, which is pretty big if you are only using it for espanso. I didn't know about "bottles", those seem exactly what we need!

I was also considering converting espanso to an App Bundle instead of a simple binary to fix some long-standing problems #208 #547 
I'm not an expert with homebrew, but [cask](https://github.com/Homebrew/homebrew-cask) seems to be the way to go in this case. Do you see any problem with this approach?

Cheers :)

**Comment by n8henrie** [comment]
I'm no homebrew expert, but I've been using it for a long time -- it's not a hard and fast rule, but a cask is generally what I would expect for a GUI app like espanso, so that's a reasonable approach.

Honestly though, one thing I really like about homebrew is having the option to build from source, ensuring that my system's security is only limited by my time, expertise, and willingness to browse through the source code. One of the nice things about a bottle is that the compilation from source is done by a trusted third party, so there is little risk of a package maintainer (or contributor) providing a malicious binary *and then validating it by providing their own hash*.

Not that my preferences should carry any weight, but opinion would be:
- the `espanso` homebrew formula should *not* be tracked in `federico-terzi/espanso` due to the circular dependency issues I noted above; homebrew formulae are already in VCS (theirs)
- a new formula can be submitted to homebrew that would contain the build-from-source option, and when a bottle is available can be updated with the appropriate hashes to allow people to skip the `rust` and `cmake` dependencies
  - I think this would be the "homebrew way" and have the best of both worlds
- the current `federico-terzi/espanso` tap and formula should be migrated to something like `federico-terzi/espanso` and `espanso-bin` respectively, so users not wanting to use the above route can still `brew tap federico-terzi/espanso` and `brew install espanso-bin`

Thanks for the prompt and thorough responses!

**Comment by n8henrie** [comment]
(On an unrelated note, I've been looking into GUI options for rust and now need to check out modulo.)

**Comment by federico-terzi** [comment]
> Honestly though, one thing I really like about homebrew is having the option to build from source, ensuring that my system's security is only limited by my time, expertise, and willingness to browse through the source code. One of the nice things about a bottle is that the compilation from source is done by a trusted third party, so there is little risk of a package maintainer (or contributor) providing a malicious binary and then validating it by providing their own hash.

I definitely understand your concerns and I agree with you. I'll make sure to move in this direction for the next version

> the espanso homebrew formula should not be tracked in federico-terzi/espanso due to the circular dependency issues I noted above; homebrew formulae are already in VCS (theirs)

There must have been a misunderstanding here because the formula resides in the https://github.com/federico-terzi/homebrew-espanso repository and not in the espanso one. So there is no circular dependency, sorry if I misinterpreted your previous message :)

> a new formula can be submitted to homebrew that would contain the build-from-source option, and when a bottle is available can be updated with the appropriate hashes to allow people to skip the rust and cmake dependencies
I think this would be the "homebrew way" and have the best of both worlds
> the current federico-terzi/espanso tap and formula should be migrated to something like federico-terzi/espanso and espanso-bin respectively, so users not wanting to use the above route can still brew tap federico-terzi/espanso and brew install espanso-bin

These are very valid points, and I think we can safely move in that direction in the future. I still want to retain a custom "tap" for homebrew to ship nightly versions fast, but major versions could be migrated to the official homebrew repository

> (On an unrelated note, I've been looking into GUI options for rust and now need to check out modulo.)

Not sure if you'll find it any useful :) It was built specifically for espanso and it can hardly be used for something else as of now

Given that this issue quickly turned around Homebrew and touched a lot of very valid points, I'm moving the M1 support thread to another issue #619

Thank you for the help!

**Comment by n8henrie** [comment]
> There must have been a misunderstanding here because the formula resides in the https://github.com/federico-terzi/homebrew-espanso repository and not in the espanso one.

Oh, interesting! I hadn't even noticed that repo, <https://github.com/federico-terzi/espanso/blob/master/packager/mac/espanso.rb> had grabbed my attention and I was focusing on it. It looks like it possibly generates the template which is copied over to the other repo. So there should be no big roadblocks, which is great!

**Comment by federico-terzi** [comment]
> It looks like it possibly generates the template which is copied over to the other repo.

Yep, that's what it does :)

**Comment by 4lch4** [comment]
Since it's being mentioned... @federico-terzi, [your article](https://federicoterzi.com/blog/how-to-publish-your-rust-project-on-homebrew/) on how you did the Homebrew Formula for Espanso actually helped me build out a Formula for the clipboard-cli app I'm so fond of.

Might be time for a v2 after you finish this all up ;) 

![image](https://user-images.githubusercontent.com/6931731/111905178-626b9d00-8a18-11eb-9855-e3b553666723.png)

**Comment by federico-terzi** [comment]
@4lch4 Yep, that might be a great time for a new article on the subject! Although we'll need to figure out how to do it first :)

**Comment by n8henrie** [comment]
Sounds good, I imagine someone will bundle up the app as a cask so we can continue to have homebrew manage it.
